Implement linked notes

kodyclemens opened the pull request:
## Description

- Notes can be hyperlinked to one another within a note body using a note's shortened UUID with the following syntax: `#h1p3u4`
- Added quick reference to a note's UUID in the note menu bar
- Clicking a note's ID will copy it to the clipboard without the hash, display 'Copied!' and return to display the ID after 3 secs

<img src="https://i.imgur.com/0ix7STZ.gif" width="450" />

- A clickable link to the referenced note is displayed in preview mode
- Clicking a link in preview mode will load the appropriate target note, replacing the currently displayed note

<img src="https://i.imgur.com/FzZRiNi.gif" width="650" />

- Invalid uuids that are attempted to be linked will show an error

<img src="https://i.imgur.com/QidtDhw.png" width="450" />

- e2e tests created for a valid and invalid note link; checks if the link is created or not based on validity of provided uuid

Closes #267 

### Future changes

- Add an overlay to the React Markdown mode to support highlighting of valid note linking syntax (again, #short-uuid-here)
- Add a select dropdown to pick a note after typing # in edit mode (like GitHub for issues)

### Browser checklist

This PR has been tested in the following browsers:

- [x] Chrome
- [ ] Firefox
- [ ] Safari

### Testing checklist

- [x] End-to-end tests have been created if necessary

taniarascia commented:
This is great, thanks for tackling it! Very nice job. A few quick comments:

- I think the note uuid "copy link to note" in the MenuBar should be simply a "Copy" icon with no text instead of "Note id: xxxxxx". The "Note copied!" notification can appear and disappear with the same timeout.
- The "copy link to note" option should also be available in the context menu. (I'm more likely to want to copy another note and paste it into the note I'm currently working on than the other way around.)
- Can we make it match something like `[#nnnnnn]` instead of just `#nnnnnn`? That way at least it looks obviously like a link from the edit mode. The way Bear app handles this is `[[note title]]` links to the Note (and two notes with the same title are considered the same note so I guess it just picks one) which might be an interesting way to go about it as well (as the user more likely wants to see the note title rather than note uuid) but either way is fine.

taniarascia commented:
I'm thinking the future idea of the select dropdown would only work if you're doing it by note title, and not by note uuid. Otherwise, you're just pasting in the exact uuid.

taniarascia commented:
So options are we can have "Copy link to note" return with:

```
[Note title](#note-uuid)
```

and that's what's used in the editor, or just use `[Note title]` and it goes to the first one it finds.

taniarascia commented on src/client/components/Editor/NoteLink.tsx:
```diff
@@ -0,0 +1,25 @@
+  const title = note !== undefined ? getNoteTitle(note.text) : null
+  const notFoundErrorMsg = '<invalid note id provided>'
+
+  if (note && title) return <a onClick={() => handleNoteLinkClick(note)}>{title}</a>
```
This needs to have `href="#!` to prevent `a` default action, or probably add `preventDefault`.

kodyclemens commented:
> * Can we make it match something like `[#nnnnnn]` instead of just `#nnnnnn`? That way at least it looks obviously like a link from the edit mode. The way Bear app handles this is `[[note title]]` links to the Note (and two notes with the same title are considered the same note so I guess it just picks one) which might be an interesting way to go about it as well (as the user more likely wants to see the note title rather than note uuid) but either way is fine.

Do you think we should use title over uuid? The uuid seems less verbose depending on the title & doesn't clutter the editor. Using uuid does remove an easy quick identification to the linked note in edit mode but it is fast to switch to preview mode to see what note is linked if you forget. Maybe it makes sense to support linking for both title and uuid? If for some reason someone has duplicate titles they could fall back to uuid, or just use uuid if they prefer it. I don't think supporting both would be any more technically challenging.

Are you okay if we just follow the standard set by Bear app and use `[[uuid]]` syntax to link notes? The idea in my head for the future feature addition was that the user, in this case, would type `[[` and be presented with a list of notes in their editor to choose from, just like GitHub does if you type `#` in this comment box. Selecting a note from the list would put in the appropriate syntax w/ uuid or title

taniarascia commented:
I think just linking to [[uuud]] would be good.

taniarascia commented:
I just don’t like the idea of typing [[ opening the GitHub-style dropdown with a list of ALL notes, because that could be hundreds. It only makes sense to me to have it be an in-line search that searches by note title.

kodyclemens commented:
- Ended up using curly braces instead of brackets to avoid unnecessary conflict with the markdown editor.
  - `{{b3n4m5}}`
- Added copy option to context menu
- Added test for context copy option
- Now using react feather icon for copy on menu bar
- Copy fn now copies the link syntax instead of just the note ID
- Added Errors enum; using to test if error message display for invalid link
- Cypress plugin config added for task `getClipboard`
- Fixed bug where if a linked note was in a different folder the active folder would not update when the note link was clicked

taniarascia commented on src/resources/LabelText.ts:
```diff
@@ -26,4 +26,5 @@ export enum LabelText {
   BACKUP_ALL_NOTES = 'Backup all notes',
   IMPORT_BACKUP = 'Import backup',
   TOGGLE_FAVORITE = 'Toggle favorite',
+  COPY_LINKED_NOTE_MARKDOWN = 'Copy linked note markdown',
```
I think this should be "Copy reference to note" .

taniarascia commented on src/client/containers/ContextMenuOptions.tsx:
```diff
@@ -191,6 +197,12 @@ const NotesOptions: React.FC<NotesOptionsProps> = ({ clickedNote }) => {
         icon={Download}
         text={LabelText.DOWNLOAD}
       />
+      <ContextMenuOption
```
The context menu should close after clicking this, that can be handled in another ticket though.

taniarascia commented on src/client/containers/NoteMenuBar.tsx:
```diff
@@ -77,6 +122,16 @@ export const NoteMenuBar = () => {
           <button className="note-menu-bar-button">
             <Download size={18} onClick={downloadNotesHandler} />
           </button>
+          <div
```
This should be a `button` like the rest of them, to retain hover and cursor effects, and instead of making the icon disappear after clicking it, just place the "Note copied!" text next to it.

taniarascia commented:
Two small comments, then it's good to go. Thanks!
